indigo-dc/onedock
# Resource allocation

## DOCKER CPU AND MEMORY ALLOCATION:

https://docs.docker.com/engine/reference/run/#runtime-constraints-on-resources
## APPROACH

The MEMORY will be assigned to control the "max limit of memory in container" in b, k, m or g. Therefore setting the "max memory of limit in the container" means that the processes in the container can use the "max limit memory in the container" assigned in the OpenNebula Templates or CLI and by default the total virtual memory size (–memory-swap) will be set as double of the "max limit of memory in the container" assigned, in this case, MEMORY + SWAP would be 2*max limit of memory in the container, so process can use the same SWAP memory as well.

The CPU will be assigned to attaching containers to cores, it means pin the container's process to a particular processor (core). It can be assigned to pin to a specific core (0), a range of cores (0-3) or some cores (1,2).
## FEATURE DESCRIPTION IN THE FIRST RELEASE

Total swap memory is set as the double of the memory specified for the container
The memory can be assigned in Megabytes by default.
The kernel must have available swap limit capabilities, otherwise the container can take as much memory as it needs (Memory limited without swap)
The CPU can be assigned for cores > 0 and for a specific core.
## HOW TO TEST
1. Install htop http://hisham.hm/htop/ to check memory and CPU workload in the host
   
   sudo apt-get install htop 
2. Create docker container attached to CPU 1 (core1)
   
   onevm create --memory 512 --cpu 1 --disk ubuntu --nic private --net_context
3. Log in into your container and install stress package, stress your container and check with htop. 
   Stress is a simple workload generator for CPU and Memory by Harvard University http://people.seas.harvard.edu/~apw/stress/
   
   apt-get install stress
   
   stress --cpu 6   (Six threads to core 1)
4. Type 'htop' command in the host to see CPU workload in CPU 1 (core1) attached previously to the container
5. Stress memory and check with htop
   
   stress --vm 1 --vm-bytes 500M --vm-hang 0
## KNOW ISSUES FOR THE FIRST RELEASE VERSION AND FUTURE WORK

Can not be assigned for CPU cores:
- The core 0
- A range of cores (0-2) 
- More than one core (0,1)

The above due to https://github.com/OpenNebula/one/blob/master/src/vm/VirtualMachine.cc#L349 , currently it does not support this kind of assignation for CPU values.

Can not be assigned memory in b, k or g due to https://github.com/OpenNebula/one/blob/master/src/vm/VirtualMachine.cc#L341 , currently it does not support "letters" assignation for memory values.

Commonly can appears a warning message as: "WARNING: Your kernel does not support swap limit capabilities, memory limited without swap", it means that limit your memory won't take a fixed limit (memory+swap) because swap is undefined, it will take as much memory as the container needs.

Be sure of the cores available in your host (number of cores) otherwise it can be crash a new container deployment, to check the number of cores in Ubuntu by example type 'grep -c processor /proc/cpuinfo'
